### PR TITLE
W-9921484 Argus Client python3 upgrade

### PR DESCRIPTION
**Pull request by mveeranala**
Argus-client code upgrade to Python 3.8.

### CODE REVIEW

File: argusclient/model.py [hunk 1]
```diff
@@ -12,19 +12,19 @@
 import json
 
 try:
-    basestring  # Python 2
+    str  # Python 2
 except NameError:
-    basestring = str  # Python 3
+    str = str  # Python 3
```
Comment (vikasm0418) [inline argusclient/model.py:17]: why is this assignment needed

**Review by vikasm0418 — APPROVED** [review]
LGTM

File: argusclient/model.py (hunk 1 above)
Comment (hdara-sfdc) [reply argusclient/model.py:17]: I see that the previous approach is the most recommended, other than using `six` package. What is the reason for this change? @mveeranala

File: argusclient/client.py [hunk 2]
```diff
@@ -511,10 +511,10 @@ def delete(self, entity_id, permission):
 
 def convert(input):
     if isinstance(input, Mapping):
-        return {convert(key): convert(value) for key, value in input.iteritems()}
+        return {convert(key): convert(value) for key, value in input.items()}
     elif isinstance(input, list):
         return [convert(element) for element in input]
-    elif isinstance(input, basestring):
+    elif isinstance(input, str):
```
Comment (hdara-sfdc) [inline argusclient/client.py:517]: This is not accurate in Python2, what is the reason behind this change?

File: argusclient/client.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 try:
     import http.client as httplib  # Python 3
 except ImportError:
-    import httplib  # Python 2
+    import http.client  # Python 2
```
Comment (hdara-sfdc) [inline argusclient/client.py:21]: How is this going to help Python 2? Won't code references to `httplib` fail?

File: examples/splunk2argus.py [hunk 4]
```diff
@@ -179,7 +179,7 @@ def get_splunk_metrics(opts):
             cols = row
             continue
         # Append rows to data
-        data.append(dict(zip(cols, row)))
+        data.append(dict(list(zip(cols, row))))
```
Comment (hdara-sfdc) [inline examples/splunk2argus.py:182]: Seems to work as is in Python 3:

```
In [142]: dict(zip([1, 2], [3, 4]))
Out[142]: {1: 3, 2: 4}
```

File: examples/splunk_to_argus.py [hunk 5]
```diff
@@ -126,7 +126,7 @@ def get_splunk_metrics(opts):
         if not cols:
             cols = row
             continue
-        data.append(dict(zip(cols, row)))
+        data.append(dict(list(zip(cols, row))))
```
Comment (hdara-sfdc) [inline examples/splunk_to_argus.py:129]: See my earlier comment.

File: setup.py [hunk 6]
```diff
@@ -34,8 +34,8 @@
       include_package_data=True,
       zip_safe=True,
       install_requires=[
-          "requests>=2.9.1",
-          "lxml>=3.2.3"
+          "requests>=2.26.0",
+          "lxml>=4.6.3"
```
Comment (hdara-sfdc) [inline setup.py:38]: Is this bump up needed specifically for Python 3

**Review by hdara-sfdc — CHANGES_REQUESTED** [review]
Why don't we start using the `six` library? It is a lot more standardized and the implementations are designed with some efficiency in mind. E.g., the unconditional `list(dict.items())` is not performant in Python 2, since it creates the list twice, instead if you use `six.iteritems()`, it does the right thing. Here are other equivalentsL
- `from six.moves import http_client`.
- `from six import string_types`
- `from six.moves import urllib`
- As mentioned before, `dict` replacements such as `iterkeys` and `itervalues`.

File: argusclient/model.py (hunk 1 above)
Comment (mveeranala) [reply argusclient/model.py]: Hari, modified it and used `six` package to address  compatibility issues.

File: argusclient/client.py (hunk 2 above)
Comment (mveeranala) [reply argusclient/client.py]: Hari, Used `String_types` from `six` package.

File: argusclient/client.py (hunk 3 above)
Comment (mveeranala) [reply argusclient/client.py]: Hari, Replaced it with `urllib` from `six.moves`

File: examples/splunk2argus.py (hunk 4 above)
Comment (mveeranala) [reply examples/splunk2argus.py]: Replaced it Hari

File: examples/splunk_to_argus.py (hunk 5 above)
Comment (mveeranala) [reply examples/splunk_to_argus.py]: Done, changed it

File: setup.py (hunk 6 above)
Comment (mveeranala) [reply setup.py:39]: Hari, in Python3 environment which is prerequisite for blackjack as mentioned in the work-item, its expecting upgraded versions of these packages. So i upgraded them to recent versions.

File: argusclient/client.py [hunk 7]
```diff
@@ -458,7 +456,7 @@ def _init_all(self, coll=None):
                                                self.get_all_req_opts.get(REQ_PATH, None),
                                                params=self.get_all_req_opts.get(REQ_PARAMS, None),
                                                dataObj=self.get_all_req_opts.get(REQ_BODY, None)))
-            for id, perms in list(resp.items()):
+            for id, perms in list(iteritems(resp)):
```
Comment (hdara-sfdc) [inline argusclient/client.py:459]:
```suggestion
            for id, perms in iteritems(resp):
```

File: argusclient/client.py [hunk 8]
```diff
@@ -477,7 +475,7 @@ def get_permissions_for_entities(self, entity_ids):
         if entityIds:
             response = convert(self.argus._request("post", "permission/entityIds", dataObj=entityIds))
             if response:
-                for id, perms in list(response.items()):
+                for id, perms in list(iteritems(response)):
```
Comment (hdara-sfdc) [inline argusclient/client.py:478]:
```suggestion
                for id, perms in iteritems(response):
```

File: argusclient/model.py [hunk 9]
```diff
@@ -10,21 +10,17 @@
...
-except NameError:
-    str = str  # Python 3
+from six import string_types, iteritems
 
 
 class BaseEncodable(object):
 
     def __init__(self, **kwargs):
-        for k, v in list(kwargs.items()):
+        for k, v in list(iteritems(kwargs)):
```
Comment (hdara-sfdc) [inline argusclient/model.py:19]:
```suggestion
        for k, v in teritems(kwargs):
```

File: argusclient/model.py [hunk 10]
```diff
@@ -10,21 +10,17 @@
...
             setattr(self, k, v)
 
     def to_dict(self):
-        D = dict((k, v) for k, v in list(self.__dict__.items()) if not k.startswith("_"))
+        D = dict((k, v) for k, v in list(iteritems(self.__dict__)) if not k.startswith("_"))
```
Comment (hdara-sfdc) [inline argusclient/model.py:23]:
```suggestion
        D = dict((k, v) for k, v in iteritems(self.__dict__) if not k.startswith("_"))
```

File: argusclient/model.py [hunk 11]
```diff
@@ -160,7 +156,7 @@ def __str__(self):
         """
         tags = hasattr(self, "tags") and self.tags or None
         metricWithTags = tags and "%s{%s}" % (
-        self.metric, ",".join("%s=%s" % (k, v) for k, v in list(self.tags.items()))) or self.metric
+        self.metric, ",".join("%s=%s" % (k, v) for k, v in list(iteritems(self.tags)))) or self.metric
```
Comment (hdara-sfdc) [inline argusclient/model.py:159]:
```suggestion
        self.metric, ",".join("%s=%s" % (k, v) for k, v in iteritems(self.tags))) or self.metric
```

File: argusclient/model.py [hunk 12]
```diff
@@ -206,7 +202,7 @@ def __str__(self):
         ``scope:metric[{tagk=tagv,...}]:source``
         """
         tags = hasattr(self, "tags") and self.tags or None
-        metricWithTags = tags and "%s{%s}" % (self.metric, ",".join("%s=%s" % (k, v) for k, v in list(self.tags.items()))) \
+        metricWithTags = tags and "%s{%s}" % (self.metric, ",".join("%s=%s" % (k, v) for k, v in list(iteritems(self.tags)))) \
```
Comment (hdara-sfdc) [inline argusclient/model.py:205]:
```suggestion
        metricWithTags = tags and "%s{%s}" % (self.metric, ",".join("%s=%s" % (k, v) for k, v in iteritems(self.tags))) \
```

**Review by hdara-sfdc — COMMENTED** [review]
Looks in general good, except for those uses in the loops that don't need a conversion to `list`.

File: argusclient/client.py [hunk 13]
```diff
@@ -206,23 +204,23 @@ def items(self):
...
+        return list(iteritems(self._coll))
 
     def keys(self):
         """
         Returns the list of ids, just like the corresponding method on a dict.
         Calling this method may result in sending a request to Argus to fetch all relevant objects.
         """
         self._init_all()
-        return list(self._coll.items())
+        return list(iteritems(self._coll))
```
[the remark this answers is not in the document]
Comment (mveeranala) [reply argusclient/client.py]: Done Hari, changed it to iterkeys.

File: argusclient/client.py (hunk 7 above)
Comment (mveeranala) [reply argusclient/client.py]: Done.

File: argusclient/client.py (hunk 8 above)
Comment (mveeranala) [reply argusclient/client.py]: Done

File: argusclient/model.py (hunk 9 above)
Comment (mveeranala) [reply argusclient/model.py]: Done

File: argusclient/model.py (hunk 10 above)
Comment (mveeranala) [reply argusclient/model.py]: Done

File: argusclient/model.py (hunk 11 above)
Comment (mveeranala) [reply argusclient/model.py]: Done

File: argusclient/model.py (hunk 12 above)
Comment (mveeranala) [reply argusclient/model.py]: Done

File: argusclient/client.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (hdara-sfdc) [reply argusclient/client.py]: Thank you.

**Review by hdara-sfdc — APPROVED** [review]
lgtm
